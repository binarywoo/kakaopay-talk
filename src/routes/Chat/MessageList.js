import React from 'react';
import Message from '../../components/Message/index';

const MessageList = ({ messages, user }) => {
  return (
    <div>
      {messages &&
        messages.map((item, idx) => {
          const isMyMessage = user.key === item.user;
          const lastUpdateMoment = window.moment(item.lastUpdate);
          const time = lastUpdateMoment.format('HH:mm');

          const prevMessage = idx > 0 ? messages[idx - 1] : null;
          const prevLastUpdateMoment = prevMessage
            ? window.moment(prevMessage.lastUpdate)
            : null;

          const prevDate = prevMessage
            ? prevLastUpdateMoment.format('YYYY-MM-DD')
            : null;
          const date = lastUpdateMoment.format('YYYY-MM-DD');

          const nextMessage =
            idx < messages.length - 1 ? messages[idx + 1] : null;
          const nextLastUpdateMoment = nextMessage
            ? window.moment(nextMessage.lastUpdate)
            : null;
          const nextTime = nextMessage
            ? nextLastUpdateMoment.format('HH:mm')
            : null;
          const nextDate = nextMessage
            ? nextLastUpdateMoment.format('YYYY-MM-DD')
            : null;
          const nextUserId =
            idx < messages.length - 1 ? nextMessage.userId : null;

          const prevUserId = prevMessage ? prevMessage.userId : null;
          const prevTime = prevMessage
            ? prevLastUpdateMoment.format('HH:mm')
            : null;

          let showTag = true;
          let showDivider = false;
          let showProfile = false;
          if (
            nextTime === time &&
            nextDate === date &&
            nextUserId === item.userId &&
            nextMessage.type !== 'participate'
          )
            showTag = false;
          if (prevDate !== date) showDivider = true;
          if (
            item.userId !== prevUserId ||
            prevTime !== time ||
            prevDate !== date ||
            prevMessage.type === 'participate'
          )
            showProfile = true;

          return (
            <Message
              key={item.key}
              isMyMessage={isMyMessage}
              showDivider={showDivider}
              showTag={showTag}
              showProfile={showProfile}
              date={date}
              time={time}
              item={item}
            />
          );
        })}
    </div>
  );
};

export default MessageList;
