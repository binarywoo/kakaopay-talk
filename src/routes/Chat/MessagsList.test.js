import React from 'react'
import { mount } from 'enzyme'
import moment from 'moment'
import MessageList from './MessageList'

describe('MessageList', () => {
  let wrapper = null
  moment.locale('ko')
  window.moment = moment
  const props = {
    messages: [
      {
        content: 'in',
        key: '-Lbx1_i6H-tlYr5-aDI2',
        lastUpdate: '2019-04-08T13:01:49.169Z',
        type: 'participate',
        user: '-Lbwm5PmrybLY2e12Mrt',
        userId: '이도희'
      },
      {
        content: '하이',
        key: '-Lbx1aoKWd_CgVTaKF43',
        lastUpdate: '2019-04-08T13:01:53.660Z',
        type: 'text',
        user: '-Lbwm5PmrybLY2e12Mrt',
        userId: '이도희'
      },
      {
        content: 'in',
        key: '-Lbx1d2RZBsTmPm8P6qF',
        lastUpdate: '2019-04-08T13:02:02.822Z',
        type: 'participate',
        user: '-Lbw8B8lmV9hY19pgs9o',
        userId: '이웅희'
      },
      {
        content: '하이',
        key: '-Lbx1eC3NVnHGd6DkaeZ',
        lastUpdate: '2019-04-08T13:02:07.532Z',
        type: 'text',
        user: '-Lbw8B8lmV9hY19pgs9o',
        userId: '이웅희'
      },
      {
        content: 'out',
        key: '-Lbx1lQl_Op0X1yXmgj-',
        lastUpdate: '2019-04-08T13:02:36.399Z',
        type: 'participate',
        user: '-Lbw8B8lmV9hY19pgs9o',
        userId: '이웅희'
      }
    ],
    user: {
      key: '-Lbwm5PmrybLY2e12Mrt',
      userId: '이도희'
    }
  }
  it('정상적으로 렌더링이 된다.', () => {
    wrapper = mount(<MessageList {...props} />)
  })

  it('스냅샷과 일치한다.', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('date divider를 정상적으로 렌더링한다.', () => {
    expect(wrapper.find('.ant-divider-inner-text').text()).toBe('2019-04-08')
  })

  it('시간 tag를 정상적으로 렌더링한다.', () => {
    expect(wrapper.find('.message-tag').length).toBe(2)
    expect(
      wrapper
        .find('.message-tag')
        .at(0)
        .text()
    ).toBe('이도희 · 22:01')
    expect(
      wrapper
        .find('.message-tag')
        .at(1)
        .text()
    ).toBe('이웅희 · 22:02')
  })

  it('참여/퇴장 메시지를 정상적으로 렌더링한다.', () => {
    expect(wrapper.find('.participate-message').length).toBe(3)
  })
})
