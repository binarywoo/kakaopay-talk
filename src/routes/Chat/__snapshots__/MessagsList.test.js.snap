// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList 스냅샷과 일치한다. 1`] = `
<MessageList
  messages={
    Array [
      Object {
        "content": "in",
        "key": "-Lbx1_i6H-tlYr5-aDI2",
        "lastUpdate": "2019-04-08T13:01:49.169Z",
        "type": "participate",
        "user": "-Lbwm5PmrybLY2e12Mrt",
        "userId": "이도희",
      },
      Object {
        "content": "하이",
        "key": "-Lbx1aoKWd_CgVTaKF43",
        "lastUpdate": "2019-04-08T13:01:53.660Z",
        "type": "text",
        "user": "-Lbwm5PmrybLY2e12Mrt",
        "userId": "이도희",
      },
      Object {
        "content": "in",
        "key": "-Lbx1d2RZBsTmPm8P6qF",
        "lastUpdate": "2019-04-08T13:02:02.822Z",
        "type": "participate",
        "user": "-Lbw8B8lmV9hY19pgs9o",
        "userId": "이웅희",
      },
      Object {
        "content": "하이",
        "key": "-Lbx1eC3NVnHGd6DkaeZ",
        "lastUpdate": "2019-04-08T13:02:07.532Z",
        "type": "text",
        "user": "-Lbw8B8lmV9hY19pgs9o",
        "userId": "이웅희",
      },
      Object {
        "content": "out",
        "key": "-Lbx1lQl_Op0X1yXmgj-",
        "lastUpdate": "2019-04-08T13:02:36.399Z",
        "type": "participate",
        "user": "-Lbw8B8lmV9hY19pgs9o",
        "userId": "이웅희",
      },
    ]
  }
  user={
    Object {
      "key": "-Lbwm5PmrybLY2e12Mrt",
      "userId": "이도희",
    }
  }
>
  <div>
    <Message
      date="2019-04-08"
      isMyMessage={true}
      item={
        Object {
          "content": "in",
          "key": "-Lbx1_i6H-tlYr5-aDI2",
          "lastUpdate": "2019-04-08T13:01:49.169Z",
          "type": "participate",
          "user": "-Lbwm5PmrybLY2e12Mrt",
          "userId": "이도희",
        }
      }
      key="-Lbx1_i6H-tlYr5-aDI2"
      showDivider={true}
      showProfile={true}
      showTag={false}
      time="22:01"
    >
      <div>
        <Divider
          className="date-divider"
        >
          <div
            className="date-divider ant-divider ant-divider-horizontal ant-divider-with-text"
          >
            <span
              className="ant-divider-inner-text"
            >
              2019-04-08
            </span>
          </div>
        </Divider>
        <div
          className="participate-message"
        >
          <b>
            이도희
          </b>
          님이
           
          대화방에 참여했습니다.
        </div>
      </div>
    </Message>
    <Message
      date="2019-04-08"
      isMyMessage={true}
      item={
        Object {
          "content": "하이",
          "key": "-Lbx1aoKWd_CgVTaKF43",
          "lastUpdate": "2019-04-08T13:01:53.660Z",
          "type": "text",
          "user": "-Lbwm5PmrybLY2e12Mrt",
          "userId": "이도희",
        }
      }
      key="-Lbx1aoKWd_CgVTaKF43"
      showDivider={false}
      showProfile={true}
      showTag={true}
      time="22:01"
    >
      <div
        className="message-row my-message"
      >
        <Alert
          className="message-box"
          message={
            <div
              dangerouslySetInnerHTML={
                Object {
                  "__html": "하이",
                }
              }
            />
          }
        >
          <Animate
            animation={Object {}}
            component=""
            componentProps={Object {}}
            onAppear={[Function]}
            onEnd={[Function]}
            onEnter={[Function]}
            onLeave={[Function]}
            showProp="data-show"
            transitionAppear={false}
            transitionEnter={true}
            transitionLeave={true}
            transitionName="ant-alert-slide-up"
          >
            <AnimateChild
              animation={Object {}}
              key="rc_animate_1554731386003"
              transitionAppear={false}
              transitionEnter={true}
              transitionLeave={true}
              transitionName="ant-alert-slide-up"
            >
              <div
                className="ant-alert ant-alert-info ant-alert-no-icon message-box"
                data-show={true}
                key="rc_animate_1554731386003"
              >
                <span
                  className="ant-alert-message"
                >
                  <div
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "하이",
                      }
                    }
                  />
                </span>
                <span
                  className="ant-alert-description"
                />
              </div>
            </AnimateChild>
          </Animate>
        </Alert>
        <div
          className="message-tag"
        >
          <span
            className="fw-700"
          >
            이도희
             · 
          </span>
          22:01
        </div>
      </div>
    </Message>
    <Message
      date="2019-04-08"
      isMyMessage={false}
      item={
        Object {
          "content": "in",
          "key": "-Lbx1d2RZBsTmPm8P6qF",
          "lastUpdate": "2019-04-08T13:02:02.822Z",
          "type": "participate",
          "user": "-Lbw8B8lmV9hY19pgs9o",
          "userId": "이웅희",
        }
      }
      key="-Lbx1d2RZBsTmPm8P6qF"
      showDivider={false}
      showProfile={true}
      showTag={false}
      time="22:02"
    >
      <div>
        <div
          className="participate-message"
        >
          <b>
            이웅희
          </b>
          님이
           
          대화방에 참여했습니다.
        </div>
      </div>
    </Message>
    <Message
      date="2019-04-08"
      isMyMessage={false}
      item={
        Object {
          "content": "하이",
          "key": "-Lbx1eC3NVnHGd6DkaeZ",
          "lastUpdate": "2019-04-08T13:02:07.532Z",
          "type": "text",
          "user": "-Lbw8B8lmV9hY19pgs9o",
          "userId": "이웅희",
        }
      }
      key="-Lbx1eC3NVnHGd6DkaeZ"
      showDivider={false}
      showProfile={true}
      showTag={true}
      time="22:02"
    >
      <div
        className="message-row other-message"
      >
        <Avatar
          icon="user"
          shape="circle"
          size="default"
          style={
            Object {
              "marginRight": 5,
              "position": "absolute",
            }
          }
        >
          <span
            className="ant-avatar ant-avatar-circle ant-avatar-icon"
            style={
              Object {
                "marginRight": 5,
                "position": "absolute",
              }
            }
          >
            <Icon
              type="user"
            >
              <LocaleReceiver
                componentName="Icon"
              >
                <i
                  aria-label="icon: user"
                  className="anticon anticon-user"
                >
                  <IconReact
                    className=""
                    type="user-o"
                  >
                    <svg
                      aria-hidden="true"
                      className=""
                      data-icon="user"
                      fill="currentColor"
                      height="1em"
                      key="svg-user"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M858.5 763.6a374 374 0 0 0-80.6-119.5 375.63 375.63 0 0 0-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 0 0-80.6 119.5A371.7 371.7 0 0 0 136 901.8a8 8 0 0 0 8 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 0 0 8-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                        key="svg-user-svg-0"
                      />
                    </svg>
                  </IconReact>
                </i>
              </LocaleReceiver>
            </Icon>
          </span>
        </Avatar>
        <Alert
          className="message-box"
          message={
            <div
              dangerouslySetInnerHTML={
                Object {
                  "__html": "하이",
                }
              }
            />
          }
        >
          <Animate
            animation={Object {}}
            component=""
            componentProps={Object {}}
            onAppear={[Function]}
            onEnd={[Function]}
            onEnter={[Function]}
            onLeave={[Function]}
            showProp="data-show"
            transitionAppear={false}
            transitionEnter={true}
            transitionLeave={true}
            transitionName="ant-alert-slide-up"
          >
            <AnimateChild
              animation={Object {}}
              key="rc_animate_1554731386003"
              transitionAppear={false}
              transitionEnter={true}
              transitionLeave={true}
              transitionName="ant-alert-slide-up"
            >
              <div
                className="ant-alert ant-alert-info ant-alert-no-icon message-box"
                data-show={true}
                key="rc_animate_1554731386003"
              >
                <span
                  className="ant-alert-message"
                >
                  <div
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "하이",
                      }
                    }
                  />
                </span>
                <span
                  className="ant-alert-description"
                />
              </div>
            </AnimateChild>
          </Animate>
        </Alert>
        <div
          className="message-tag"
        >
          <span
            className="fw-700"
          >
            이웅희
             · 
          </span>
          22:02
        </div>
      </div>
    </Message>
    <Message
      date="2019-04-08"
      isMyMessage={false}
      item={
        Object {
          "content": "out",
          "key": "-Lbx1lQl_Op0X1yXmgj-",
          "lastUpdate": "2019-04-08T13:02:36.399Z",
          "type": "participate",
          "user": "-Lbw8B8lmV9hY19pgs9o",
          "userId": "이웅희",
        }
      }
      key="-Lbx1lQl_Op0X1yXmgj-"
      showDivider={false}
      showProfile={false}
      showTag={true}
      time="22:02"
    >
      <div>
        <div
          className="participate-message"
        >
          <b>
            이웅희
          </b>
          님이
           
          대화방을 나갔습니다.
        </div>
      </div>
    </Message>
  </div>
</MessageList>
`;
